clients/vault/src/oracle/collector/proof_builder.rs

```diff
@@ -1,160 +1,261 @@
+					}
+					// Use Horizon Archive node to get ScpHistoryEntry
+					return Err(ProofStatus::LackingEnvelopes)
+				}
 
-		let mut vec_envelopes = vec![];
+				Ok(UnlimitedVarArray::new(envelopes.clone())
+					.unwrap_or(UnlimitedVarArray::new_empty()))
+			},
+		}
+	}
 
-		let fetch_more = if envelopes.is_none() {
-			true
-		} else {
-			vec_envelopes = envelopes.unwrap().0;
-			vec_envelopes.len() < get_min_externalized_messages(self.is_public())
-		};
+	/// Returns either a TransactionSet or a ProofStatus saying it failed to retrieve the set.
+	fn get_txset(&self, slot: Slot) -> Result<TransactionSet, ProofStatus> {
+		match self.txset_map().get(&slot).map(|set| set.clone()) {
+			None => {
+				// If the current slot is still in the range of 'remembered' slots
+				if check_slot_position(*self.last_slot_index(), slot) {
+					return Err(ProofStatus::WaitForTxSet)
+				}
 
-		if fetch_more {
-			self.restore_missed_slots(slot);
-			return Err(ProofStatus::LackingEnvelopes(slot))
+				Err(ProofStatus::NoTxSetFound)
+			},
+			Some(res) => Ok(res),
 		}
-
-		Ok(UnlimitedVarArray::new(vec_envelopes).unwrap_or(UnlimitedVarArray::new_empty()))
 	}
 
-	fn restore_missed_slots(&self, slot: Slot) {
-		let last_slot_index = *self.last_slot_index();
-		let action_sender = self.action_sender.clone();
-		let rw_lock = self.envelopes_map_clone();
-		tokio::spawn(async move {
-			// If the current slot is still in the range of 'remembered' slots
-			if slot > last_slot_index - MAX_SLOT_TO_REMEMBER {
-				let result =
-					action_sender.send(ActorMessage::GetScpState { missed_slot: slot }).await;
-			} else {
-				let slot_index: u32 = slot.try_into().unwrap();
-				let scp_archive: XdrArchive<ScpHistoryEntry> =
-					ScpArchiveStorage::get_scp_archive(slot.try_into().unwrap()).await.unwrap();
-
-				let value = scp_archive.get_vec().into_iter().find(|&scp_entry| {
-					if let ScpHistoryEntry::V0(scp_entry_v0) = scp_entry {
-						return scp_entry_v0.ledger_messages.ledger_seq == slot_index
-					} else {
-						return false
-					}
-				});
-
-				if let Some(i) = value {
-					if let ScpHistoryEntry::V0(scp_entry_v0) = i {
-						let slot_scp_envelopes = scp_entry_v0.clone().ledger_messages.messages;
-						let vec_scp = slot_scp_envelopes.get_vec().clone(); //TODO store envelopes_map or send via mpsc
+	/// Returns a `ProofStatus`.
+	///
+	/// # Arguments
+	///
+	/// * `slot` - The slot of a transactionwe want a proof of.
+	/// * `overlay_conn` - The StellarOverlayConnection used for sending messages to Stellar Node
+	pub async fn build_proof(
+		&mut self,
+		slot: Slot,
+		overlay_conn: &StellarOverlayConnection,
+	) -> ProofStatus {
+		// get the SCPEnvelopes
+		let envelopes = match self.get_envelopes(slot) {
+			Ok(envelopes) => envelopes,
+			Err(neg_status) => {
+				match &neg_status {
+					// let's fetch from Stellar Node again
+					ProofStatus::WaitForMoreEnvelopes | ProofStatus::WaitForEnvelopes => {
```

Comment:
Do we really need separate types for this? Maybe just having `WaitForEnvelopes` is enough.

Reply:
removed those extras variants.